test/clj/matcher_combinators/core_test.clj

```diff
@@ -572,4 +579,36 @@
                                   :in-any-order)
             ::result/weight 2}))
 
+(deftest matcher-for-special-cases
+  (testing "matcher for a fn is a fn"
+    (is (fn? (core/matcher-for (fn [])))))
```

Comment:
I feel like this is a little misleading. As far as I recall, the fact that it is a function is just because of how we do dispatch for function matching.
For instance, it could (and maybe should) be reimplemented like this:

```
diff --git a/src/cljc/matcher_combinators/dispatch.cljc b/src/cljc/matcher_combinators/dispatch.cljc
index c4f3d65..4c475ba 100644
--- a/src/cljc/matcher_combinators/dispatch.cljc
+++ b/src/cljc/matcher_combinators/dispatch.cljc
@@ -58,7 +58,8 @@
 
 ;; other
 (defn pattern-dispatch [expected] (matchers/regex expected))
-(defn function-dispatch [expected] (partial core/match-pred expected (str "predicate: " expected)))
+(defn function-dispatch [expected] (core/->PredMatcher expected
+                                                       (str "predicate: " expected)))
 
 (def type->dispatch
   #?(:cljs {}
diff --git a/src/cljc/matcher_combinators/parser.cljc b/src/cljc/matcher_combinators/parser.cljc
index 58c5454..5c54aeb 100644
--- a/src/cljc/matcher_combinators/parser.cljc
+++ b/src/cljc/matcher_combinators/parser.cljc
@@ -18,7 +18,7 @@
   ;; function as predicate
   function
   (match [this actual]
-    ((dispatch/function-dispatch this) actual))
+    (core/match (dispatch/function-dispatch this) actual))
 
   ;; equals base types
   nil
@@ -107,7 +107,7 @@
   core/Matcher
   (matcher-for [this] (dispatch/function-dispatch this))
   (match [this actual]
-    ((dispatch/function-dispatch this) actual)))
+    (core/match (dispatch/function-dispatch this) actual)))
 
 (mimic-matcher dispatch/nil-dispatch nil)
 (mimic-matcher dispatch/class-dispatch java.lang.Class)
```

Reply:
I merged this patch and also modified it further, in a separate commit, to go through a matcher fn to align with the other dispatch fns.